### PR TITLE
core/types: transaction and receipt encoding/decoding optimizations

### PR DESCRIPTION
**Pull request by fjl**
Just some minor optimizations I figured out a while ago. By using `ReadBytes` instead of
`Bytes` on the rlp stream, we can save the allocation of a temporary buffer for the typed tx
payload.

### CODE REVIEW

**Comment by fjl** [comment]
I'm not really sure about these benchmark results. Will rerun on another machine to verify.

**Comment by fjl** [comment]
Here are some more stable results. It looks like the effect of this change is very minimal:

```
goos: freebsd
goarch: amd64
pkg: github.com/ethereum/go-ethereum/core/types
cpu: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
                                │ types.old.txt │            types.new.txt            │
                                │    sec/op     │    sec/op     vs base               │
EncodeRLP/legacy-header-6          242.1n ±  1%   239.2n ±  1%  -1.24% (p=0.000 n=10)
EncodeRLP/london-header-6          252.7n ±  1%   250.6n ±  1%  -0.81% (p=0.006 n=10)
EncodeRLP/receipt-for-storage-6    128.9n ±  0%   124.3n ±  0%  -3.64% (p=0.000 n=10)
EncodeRLP/receipt-full-6           303.6n ±  0%   305.3n ±  1%  +0.54% (p=0.000 n=10)
EncodeRLP/legacy-transaction-6     405.3n ±  1%   379.9n ±  0%  -6.27% (p=0.000 n=10)
EncodeRLP/access-transaction-6     607.3n ±  0%   590.9n ±  0%  -2.71% (p=0.000 n=10)
EncodeRLP/1559-transaction-6       591.8n ±  0%   581.2n ±  0%  -1.79% (p=0.000 n=10)
DecodeRLP/legacy-header-6          931.0n ±  0%   930.0n ±  0%  -0.12% (p=0.000 n=10)
DecodeRLP/london-header-6          953.2n ±  0%   950.5n ±  0%  -0.28% (p=0.000 n=10)
DecodeRLP/receipt-for-storage-6    506.2n ±  4%   514.0n ±  5%       ~ (p=0.353 n=10)
DecodeRLP/receipt-full-6           585.3n ±  2%   586.6n ±  4%  +0.23% (p=0.010 n=10)
DecodeRLP/legacy-transaction-6     1.278µ ± 10%   1.252µ ± 13%       ~ (p=0.869 n=10)
DecodeRLP/access-transaction-6     1.507µ ±  9%   1.555µ ±  5%       ~ (p=0.564 n=10)
DecodeRLP/1559-transaction-6       1.710µ ±  9%   1.635µ ±  9%       ~ (p=0.325 n=10)
geomean                            560.5n         553.2n        -1.30%

                                │ types.old.txt │            types.new.txt             │
                                │      B/s      │      B/s       vs base               │
EncodeRLP/legacy-header-6          2.089Gi ± 1%   2.115Gi ±  1%  +1.25% (p=0.000 n=10)
EncodeRLP/london-header-6          2.023Gi ± 1%   2.040Gi ±  1%  +0.81% (p=0.007 n=10)
EncodeRLP/receipt-for-storage-6    66.55Mi ± 0%   69.10Mi ±  0%  +3.83% (p=0.000 n=10)
EncodeRLP/receipt-full-6           848.1Mi ± 0%   843.5Mi ±  1%  -0.53% (p=0.000 n=10)
EncodeRLP/legacy-transaction-6     240.0Mi ± 1%   256.0Mi ±  0%  +6.69% (p=0.000 n=10)
EncodeRLP/access-transaction-6     166.4Mi ± 0%   171.1Mi ±  0%  +2.79% (p=0.000 n=10)
EncodeRLP/1559-transaction-6       177.3Mi ± 0%   180.5Mi ±  0%  +1.82% (p=0.000 n=10)
DecodeRLP/legacy-header-6          556.2Mi ± 0%   556.9Mi ±  0%  +0.12% (p=0.000 n=10)
DecodeRLP/london-header-6          549.3Mi ± 0%   550.8Mi ±  0%  +0.28% (p=0.000 n=10)
DecodeRLP/receipt-for-storage-6    16.96Mi ± 3%   16.70Mi ±  5%       ~ (p=0.362 n=10)
DecodeRLP/receipt-full-6           439.9Mi ± 2%   438.9Mi ±  4%  -0.23% (p=0.011 n=10)
DecodeRLP/legacy-transaction-6     76.14Mi ± 9%   77.70Mi ± 11%       ~ (p=0.912 n=10)
DecodeRLP/access-transaction-6     67.10Mi ± 8%   64.99Mi ±  4%       ~ (p=0.579 n=10)
DecodeRLP/1559-transaction-6       61.39Mi ± 8%   64.18Mi ±  8%       ~ (p=0.325 n=10)
geomean                            231.4Mi        234.5Mi        +1.31%

                                │ types.old.txt │            types.new.txt             │
                                │     B/op      │    B/op     vs base                  │
EncodeRLP/legacy-header-6          0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
EncodeRLP/london-header-6          0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
EncodeRLP/receipt-for-storage-6    0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
EncodeRLP/receipt-full-6           320.0 ± 0%     320.0 ± 0%        ~ (p=1.000 n=10) ¹
EncodeRLP/legacy-transaction-6     0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
EncodeRLP/access-transaction-6     24.00 ± 0%     24.00 ± 0%        ~ (p=1.000 n=10) ¹
EncodeRLP/1559-transaction-6       24.00 ± 0%     24.00 ± 0%        ~ (p=1.000 n=10) ¹
DecodeRLP/legacy-header-6          112.0 ± 0%     112.0 ± 0%        ~ (p=1.000 n=10) ¹
DecodeRLP/london-header-6          112.0 ± 0%     112.0 ± 0%        ~ (p=1.000 n=10) ¹
DecodeRLP/receipt-for-storage-6    144.0 ± 0%     144.0 ± 0%        ~ (p=1.000 n=10) ¹
DecodeRLP/receipt-full-6           393.0 ± 0%     393.0 ± 0%        ~ (p=1.000 n=10) ¹
DecodeRLP/legacy-transaction-6     480.0 ± 0%     480.0 ± 0%        ~ (p=1.000 n=10) ¹
DecodeRLP/access-transaction-6     736.0 ± 0%     624.0 ± 0%  -15.22% (p=0.000 n=10)
DecodeRLP/1559-transaction-6       768.0 ± 0%     656.0 ± 0%  -14.58% (p=0.000 n=10)
geomean                                       ²                -2.28%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                │ types.old.txt │            types.new.txt            │
                                │   allocs/op   │ allocs/op   vs base                 │
EncodeRLP/legacy-header-6          0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
EncodeRLP/london-header-6          0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
EncodeRLP/receipt-for-storage-6    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
EncodeRLP/receipt-full-6           1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
EncodeRLP/legacy-transaction-6     0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
EncodeRLP/access-transaction-6     1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
EncodeRLP/1559-transaction-6       1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
DecodeRLP/legacy-header-6          3.000 ± 0%     3.000 ± 0%       ~ (p=1.000 n=10) ¹
DecodeRLP/london-header-6          3.000 ± 0%     3.000 ± 0%       ~ (p=1.000 n=10) ¹
DecodeRLP/receipt-for-storage-6    5.000 ± 0%     5.000 ± 0%       ~ (p=1.000 n=10) ¹
DecodeRLP/receipt-full-6           4.000 ± 0%     4.000 ± 0%       ~ (p=1.000 n=10) ¹
DecodeRLP/legacy-transaction-6     13.00 ± 0%     13.00 ± 0%       ~ (p=1.000 n=10) ¹
DecodeRLP/access-transaction-6     18.00 ± 0%     17.00 ± 0%  -5.56% (p=0.000 n=10)
DecodeRLP/1559-transaction-6       19.00 ± 0%     18.00 ± 0%  -5.26% (p=0.000 n=10)
geomean                                       ²               -0.79%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

File: core/types/hashing.go [hunk 1]
```diff
@@ -36,6 +38,21 @@ var encodeBufferPool = sync.Pool{
 	New: func() interface{} { return new(bytes.Buffer) },
 }
 
+// getPooledBuffer retrieves a buffer from the pool and creates a byte slice
+// of the requested size from it.
+//
+// The caller should return the Buffer object back into encodeBufferPool after use!
```
Comment (holiman) [inline core/types/hashing.go:44]: Would be good to clarify the semantics for the byte slice too, afaiu it also must not be kept after return? 
Something like this? 
```suggestion
// The caller should return the Buffer object back into encodeBufferPool after use, and must release the slice!
```

**Comment by fjl** [comment]
This should be ready now. The package rlp change has moved to #27987

File: core/types/receipt.go [hunk 2]
```diff
@@ -264,8 +270,8 @@ func (r *ReceiptForStorage) EncodeRLP(_w io.Writer) error {
 	w.WriteUint64(r.CumulativeGasUsed)
 	logList := w.List()
 	for _, log := range r.Logs {
-		if err := rlp.Encode(w, log); err != nil {
-			return err
+		if err := log.EncodeRLP(w); err != nil {
+			return nil
```
Comment (jsvisa) [inline core/types/receipt.go:274]: this error is ignored?

Comment (fjl) [reply core/types/receipt.go:274]: Good catch!

Comment (holiman) [inline core/types/receipt.go:274]:
```suggestion
			return err
```

**Pull request merged**